op-program/client/l2/fast_canon.go

```diff
@@ -0,0 +1,165 @@
+package l2
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus"
+	"github.com/ethereum/go-ethereum/consensus/beacon"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/triedb"
+	"github.com/hashicorp/golang-lru/v2/simplelru"
+)
+
+// historicalCacheSize is the number of cached eip-2935 historical block lookups
+// This covers 4 weeks worth of blocks on a 2 second block time.
+// We keep the cache size small to reduce memory usage and to ensure cache scans are fast.
+const historicalCacheSize = 160
+
+type FastCanonicalBlockHeaderOracle struct {
+	head          *types.Header
+	blockByHashFn BlockByHashFn
+	config        *params.ChainConfig
+	fallback      *CanonicalBlockHeaderOracle
+	ctx           *chainContext
+	db            ethdb.KeyValueStore
+	cache         *simplelru.LRU[uint64, *types.Header]
+}
+
+func NewFastCanonicalBlockHeaderOracle(
+	head *types.Header,
+	blockByHashFn BlockByHashFn,
+	chainCfg *params.ChainConfig,
+	stateOracle StateOracle,
+	kvdb KeyValueStore,
+	fallback *CanonicalBlockHeaderOracle,
+) *FastCanonicalBlockHeaderOracle {
+	chainID := eth.ChainIDFromBig(chainCfg.ChainID)
+	ctx := &chainContext{engine: beacon.New(nil)}
+	db := NewOracleBackedDB(kvdb, stateOracle, chainID)
+	cache, _ := simplelru.NewLRU[uint64, *types.Header](historicalCacheSize, nil)
+	return &FastCanonicalBlockHeaderOracle{
+		head:          head,
+		blockByHashFn: blockByHashFn,
+		config:        chainCfg,
+		fallback:      fallback,
+		ctx:           ctx,
+		db:            db,
+		cache:         cache,
+	}
+}
+
+func (o *FastCanonicalBlockHeaderOracle) CurrentHeader() *types.Header {
+	return o.head
+}
+
+func (o *FastCanonicalBlockHeaderOracle) GetHeaderByNumber(n uint64) *types.Header {
+	if o.head.Number.Uint64() < n {
+		return nil
+	}
+	if o.head.Number.Uint64() == n {
+		return o.head
+	}
+
+	// scan the cache for a header that contains the requested block in its historical block window
+	cover := uint64(math.MaxUint64)
+	for _, number := range o.cache.Keys() {
+		if number >= n && number < cover {
+			cover = number
+		}
+	}
+	h := o.head
+	if cover != math.MaxUint64 {
+		h, _ = o.cache.Get(cover)
+	}
+
+	for h.Number.Uint64() >= n {
+		headNumber := h.Number.Uint64()
+		if headNumber == n {
+			return h
+		}
```

Comment:
Any reason we don't just make this:
```suggestion
	for h.Number.Uint64() > n {
		headNumber := h.Number.Uint64()
```

Reply:
No reason. I've implemented your suggestion.